Made Split to Nearby Query into a fast query for coordinates + a details query for each pin

kanahia1 opened the pull request:
**Description (required)**

Fixes #4560 

What changes did you make and why?
- Splitted to the query into two queries
- Improved the pop up for the marker which will probably fix https://github.com/commons-app/apps-android-commons/issues/5692
- https://github.com/commons-app/apps-android-commons/issues/5664

**Tests performed (required)**

Tested prodDebug on Samsung S21 FE with API level 33.

**Screenshots (for UI changes only)**

https://github.com/commons-app/apps-android-commons/assets/114223204/fac1c647-b014-4275-8c56-6ce82e07e117

nicolas-raoul commented:
Testing now.

Just wondering: Is there a specific reason why your branch is called `kanahia1-main`?
It is usually good to have a branch name that reflects the issue ID and an extremely short summary of the issue.
It makes it easier to work on different things in parallel. :-)

kanahia1 commented:
There's no specific reason for it 😄 From next time I will try to follow proper naming.

nicolas-raoul commented on app/src/main/resources/queries/rectangle_query_for_nearby.rq:
```diff
@@ -1,16 +1,9 @@
   (SAMPLE(?location) as ?location)
-  (SAMPLE(?label) AS ?label)
-  (SAMPLE(?description) AS ?description)
   (SAMPLE(?class) AS ?class)
```
Could you please move this to the second query?

kanahia1 replied:
For showing the markers on the map we will require class, endtime and destroyed. As class is used for the filtering type such as building, church  etc. Which we utilize to show different icons for the pins. And endtime, destroyed are required for the filters such as Exists.

nicolas-raoul replied:
I believe these properties make the query slower (the query took about 10 seconds when I tried now).
Filtering is not as important as showing the first points as fast as possible.

As stated in the issue, the first query must really be only for coordinates.

Then as the points' details are loaded little by little, the points that should not be shown will be removed from the map according to the filters such as `Needs Photo` (which currently does not have any effect).

While doing tests, I recommend using FakeGPS or similar to put your location in a busy place such as the South of Manhattan. :-)

kanahia1 replied:
Oh, I got the idea!

nicolas-raoul commented on app/src/main/resources/queries/rectangle_query_for_nearby.rq:
```diff
@@ -1,16 +1,9 @@
-  (SAMPLE(?classLabel) AS ?classLabel)
-  (SAMPLE(?pic) AS ?pic)
   (SAMPLE(?destroyed) AS ?destroyed)
   (SAMPLE(?endTime) AS ?endTime)
```
Could you please move these two to the second query?

nicolas-raoul reviewed (commented):
Would you mind implementing a loop that asynchrously loads each item's details, for all shown points? Depending on the user's settings it will make many points disappear, which is not ideal user experience but it is the best we can do while having fast nearby maps in areas with many items.

Thanks! 🙂

kanahia1 commented:
Sure, I will do asynchronous loop to load pin's details

nicolas-raoul commented:
Sounds good! :-)

For the second query, you can specify the QIDs of the pins, that will make the query much faster.
Here is a example of how to specify QIDs: https://www.wikidata.org/wiki/User:Vojt%C4%9Bch_Dost%C3%A1l/app-Czech_uploads
I believe this is the relevant documentation: https://www.w3.org/TR/sparql11-query/#inline-data
Obviously, the second query does not need to care about coordinates.
Let me know if anything is unclear. :-)

nicolas-raoul commented:
By the way, I believe it would be user-friendly to first show pins in grey color to mean "unfiltered". Then upon receiving the result from the second request, hide the pins that do not match the filter, and set the normal color to the pins that match the filter. What do you think? 🙂

kanahia1 commented:
Yeah, That would be great!

kanahia1 commented:
Hello @nicolas-raoul, I have improvised query as per suggestions by you and it's working fine for the smaller regions

https://github.com/commons-app/apps-android-commons/assets/114223204/9e173ad8-fc16-4a7f-8433-44e74ce2cdab

But I have faced an issue because we are using List of QID (which becomes really large for city like capital of India) the server returns error 414 which is because our url length exceeds a limit (which is because of the large query)

https://github.com/commons-app/apps-android-commons/assets/114223204/da7a8609-3214-433d-98f3-e143c2aa0759

How about replacing the List of QID with similar method we had earlier?

nicolas-raoul commented:
Would you mind splitting the second query into several queries of 50 QIDs max, and using the results as they come? Ideally sort them by their distance to the center of the search rectangle, then run the queries starting from the nearest 50 QIDs.

Sorry to add something again, but when a not-yet-loaded QID is tapped by the user, it would be great to run the second query for that point, in parallel with the above, it will probably return the result fast and the user will see the label/class/filter status/etc without having to wait for all points to load.

kanahia1 commented:
Thank you for suggestion and I feel it will make queries even better.

And I have added that when pins are not loaded still the user can get the details for a pin.

https://github.com/commons-app/apps-android-commons/assets/114223204/9113a593-a5a4-41fb-a9c1-b1ece53cfde2

kanahia1 commented:
Hello Nicolas, I have divided the query into batches fixes the issue we were facing. Now the first query has became more effective.

https://github.com/commons-app/apps-android-commons/assets/114223204/61548181-304c-439f-a51a-fd80f170feac

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -1354,15 +1399,72 @@ private void populatePlacesForAnotherLocation(
 
+    @SuppressLint("CheckResult")
+    public void loadPlacesDataAsync(List<Place> placeList, LatLng curLatLng) {
+        List<Place> places = new ArrayList<>(placeList);
```
It would be great to sort the list here, in order of growing distance to curLatLng.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:
```diff
@@ -397,6 +399,82 @@ public List<Place> getNearbyPlaces(
+     * @throws Exception If an error occurs during the retrieval process.
+     */
+    @Nullable
+    public Place getNearbyPlace(
```
How about renaming to `getPlace(`?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:
```diff
@@ -397,6 +399,82 @@ public List<Place> getNearbyPlaces(
+    }
+
+    @Nullable
+    public List<Place> getNearbyPlacesFromQID(
```
How about renaming to `getPlaces(`?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java:
```diff
@@ -120,6 +120,26 @@ public List<Place> getFromWikidataQuery(
+     * @return A place obtained from the Wikidata query.
+     * @throws Exception If an error occurs during the retrieval process.
+     */
+    public Place getPlaceFromWikidataQuery(final String entityId,
```
How about renaming to `getPlace(`?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java:
```diff
@@ -120,6 +120,26 @@ public List<Place> getFromWikidataQuery(
+            .getNearbyPlace(entityId, lang);
+    }
+
+    public List<Place> getPlacesFromQID(final List<Place> placeList,
```
How about renaming to `getPlaces(`?

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -1354,15 +1399,72 @@ private void populatePlacesForAnotherLocation(
+    public void loadPlacesDataAsync(List<Place> placeList, LatLng curLatLng) {
+        List<Place> places = new ArrayList<>(placeList);
+        int batchSize = 150;
+        List<Place> processedPlaces = new ArrayList<>();
```
With the current implementation, too many requests will be sent to the Wikimedia server, making it send HTTP error 429.
So, the loop should wait for each request to get its results before doing the next request.
But refreshing the map can be asynchronous.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -1354,15 +1399,72 @@ private void populatePlacesForAnotherLocation(
+    @SuppressLint("CheckResult")
+    public void loadPlacesDataAsync(List<Place> placeList, LatLng curLatLng) {
+        List<Place> places = new ArrayList<>(placeList);
+        int batchSize = 150;
```
Judging from your screencast 150 seems to take a lot of time (actually I am getting `Could not load nearby places - Unexpected response code: 500`.), even in the countryside for 5 points it feels slow, so how about `3` for now? Thanks :-)

nicolas-raoul reviewed (changes requested):
It is getting better and better, keep up the great work :-)
Brilliant work on the SPARQL query especially!

kanahia1 commented:
Hello Nicolas, Thanks for the suggestions and now loading of pins has became even more efficient.

https://github.com/commons-app/apps-android-commons/assets/114223204/85b9d03e-43fc-4ef7-8377-10181f243f6f

nicolas-raoul commented:
Looks fantastic! I will try it as soon as I get home. 🙂

kanahia1 commented:
Hi @nicolas-raoul, on digging I am able to find that issue was caused by ongoing network call which was not stopped on starting new call. Now I have updated the code, I hope it should fix up the issue. 

https://github.com/commons-app/apps-android-commons/assets/114223204/d813f09f-71ef-4b43-ae6f-7e5cbf5d29da

nicolas-raoul commented:
I am experiencing several unexpected things, especially when swgitching to other apps or when the screen goes black:

https://github.com/commons-app/apps-android-commons/assets/99590/e81e25cf-5089-478f-a0a5-9f37f46f45f0

Also, maybe modifying the default zoom level would be good (zoomed more, mening it loads less items by default).

kanahia1 commented:
Sure, I will update the zoom level ;-)

I tried switching apps. It moves the map to the default location 

https://github.com/commons-app/apps-android-commons/assets/114223204/0042658f-f765-4e87-8340-e8d36e7f007d

Actually It wasn't clear from the video provided. On switching apps the map return to the my location (in my case)

nicolas-raoul commented:
Let's say the map is fully loaded, including points details (exist or not, etc). Then when I turn the screen off and on again, all points on the map turned back to grey again, and it starts sending requests again 3-by-3. It should not reload unless it is kicked ou of memory (this happens on a device with lots of memory and no other running app so memory is not the issue).
I send a screencast on Zulip.

kanahia1 commented:
Hey Nicolas, I have fixed the unnecessary reloads caused onResume. 

https://github.com/commons-app/apps-android-commons/assets/114223204/d8dc71ab-a202-45b3-b5b2-158c9f7eb2a7

We may also need to implement method to load pending pins which are not loaded on changing the apps as it returns <-- HTTP FAILED: java.io.InterruptedIOException, I am working over it and will also share a commit for it ASAP

nicolas-raoul commented:
Great, that problem is now fixed. :-)

Another issue:

when moving and searching another area, the pins details load from the side, rather than from the center. Because the side is often hidden, it sometimes looks like the details are not loading.

Also, any chance to modify the default zoom level? :-)

nicolas-raoul commented:
I don't think we should limit the zoom level. Just start at a different zoom level when the Nearby activity starts. Currently it always starts at this scale, which is too wide for most urban areas:
![Screenshot_20240530-070637.png](https://github.com/commons-app/apps-android-commons/assets/99590/212f16b9-5827-443f-a351-a9f0df09c487)

nicolas-raoul commented:
Somehow https://wikishootme.toolforge.org/#lat=35.657885840709675&lng=139.70094680786136&zoom=14&layers=wikidata_image,wikidata_no_image is abou 100 times faster than our first query, I wonder what's different.

nicolas-raoul commented:
Somehow https://wikishootme.toolforge.org/#lat=35.657885840709675&lng=139.70094680786136&zoom=14&layers=wikidata_image,wikidata_no_image is abou 100 times faster than our first query, I wonder what's different.

kanahia1 commented:
Hey Nicolas, I have set the default zoom level to 0.5KM and have also fixed the pins not loading from the center.

https://github.com/commons-app/apps-android-commons/assets/114223204/fb2a3212-7442-45ef-83b9-1354888c5b2f

I am looking into the implementation by wikishootme and will share the details.

kanahia1 commented:
I looked into the WikiShootMe implementation and here's what I have found.

**Request URL for the query by Wikishootme** 
https://query.wikidata.org/bigdata/namespace/wdq/sparql?query=%23TOOL%3A%20WikiShootMe%0ASELECT%20%3Fq%20%3FqLabel%20%3Flocation%20%3Fimage%20%3Freason%20%3Fdesc%20%3Fcommonscat%20%3Fstreet%20WHERE%20%7B%20SERVICE%20wikibase%3Abox%20%7B%20%3Fq%20wdt%3AP625%20%3Flocation%20.%20bd%3AserviceParam%20wikibase%3AcornerSouthWest%20%22Point(139.6627521514893%2035.63406978523804)%22%5E%5Egeo%3AwktLiteral%20.%20bd%3AserviceParam%20wikibase%3AcornerNorthEast%20%22Point(139.74635124206546%2035.68163144949355)%22%5E%5Egeo%3AwktLiteral%20%7D%20%20OPTIONAL%20%7B%20%3Fq%20wdt%3AP18%20%3Fimage%20%7D%20%20OPTIONAL%20%7B%20%3Fq%20wdt%3AP373%20%3Fcommonscat%20%7D%20%20OPTIONAL%20%7B%20%3Fq%20wdt%3AP969%20%3Fstreet%20%7D%20%20SERVICE%20wikibase%3Alabel%20%7B%20bd%3AserviceParam%20wikibase%3Alanguage%20%22en%2Cen%2Cde%2Cfr%2Ces%2Cit%2Cnl%22%20.%20%3Fq%20schema%3Adescription%20%3Fdesc%20.%20%3Fq%20rdfs%3Alabel%20%3FqLabel%20%7D%20%7D%20LIMIT%203000

![Screenshot 2024-05-30 095632](https://github.com/commons-app/apps-android-commons/assets/114223204/a6881064-aca8-40f5-8468-8586b5487b79)

**Request URL for the same point by the App** 
https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%28139.6627521514893%2035.63406978523804%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%28139.74635124206546%2035.68163144949355%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json

![Screenshot 2024-05-30 095642](https://github.com/commons-app/apps-android-commons/assets/114223204/28b112f6-6318-4242-b796-bb76e7bd068f)

_Even though the wikishootme requests double data than the commons app query (including desc, image, categories) it is able to get data in the same amount of time._

I thought it may be due to url used by the wikishootme which is `https://query.wikidata.org/bigdata/namespace/wdq/sparql` instead of `https://query.wikidata.org/sparql`

Code used for calculating response time -
```
        long startTime = System.nanoTime(); // Start time
        final Response response = okHttpClient.newCall(request).execute();
        long endTime = System.nanoTime(); // End time
        long durationInMillis = (endTime - startTime) / 1_000_000; // Calculate duration in milliseconds
```

Points - 77.2100 28.6203, 77.2299 28.6433

**Results with URL https://query.wikidata.org/sparql**

```
https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2100%2028.6203%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2299%2028.6433%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 1020 ms

https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2102%2028.6199%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2301%2028.6430%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 975 ms

https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2097%2028.6193%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2296%2028.6424%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 722 ms
```

**Results with URL https://query.wikidata.org/bigdata/namespace/wdq/sparql**

```
https://query.wikidata.org/bigdata/namespace/wdq/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2092%2028.6203%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2294%2028.6436%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 700 ms

https://query.wikidata.org/bigdata/namespace/wdq/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2102%2028.6198%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2303%2028.6431%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 879 ms

https://query.wikidata.org/bigdata/namespace/wdq/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flocation%29%20as%20%3Flocation%29%0AWHERE%20%7B%0A%20%20%23%20Around%20given%20location%0A%20%20SERVICE%20wikibase%3Abox%20%7B%0A%20%20%20%20%3Fitem%20wdt%3AP625%20%3Flocation.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerWest%20%22Point%2877.2093%2028.6211%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%20%20%20bd%3AserviceParam%20wikibase%3AcornerEast%20%22Point%2877.2294%2028.6445%29%22%5E%5Egeo%3AwktLiteral.%0A%20%20%7D%0A%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
HTTP request response time: 710 ms
```

nicolas-raoul commented:
Thanks a lot for the in-depth research!!! :-)

nicolas-raoul commented:
So https://query.wikidata.org/sparql and https://query.wikidata.org/bigdata/namespace/wdq/sparql seem to have rather similar speeds, right?

I see the Wikishootme request does not specify `&format=json` and gets a .srx response.

Improbable but maybe the server-side srx->JSON transformation is particularly slow sometimes? If that's the case, let's create a separate issue about switching Nearby to .srx hopefully someone else might be interested in implementing it.

I will try your new commits now. :-)

kanahia1 commented:
That could also be a case, will also look into it.

kanahia1 commented:
I have changed the pin's color and removed toggle chips https://github.com/commons-app/apps-android-commons/pull/5731/commits/c56a16e38e4dd6e16a2d94c1e629fbd70a112a61

![WhatsApp Image 2024-06-02 at 09 41 04_0c80e5f7](https://github.com/commons-app/apps-android-commons/assets/114223204/cc422267-5610-40df-8c8c-005c4240d417)

nicolas-raoul commented:
Would you mind solving the conflict? :-)

nicolas-raoul commented:
One point stays grey strangely:
![Screenshot_20240612-215632.png](https://github.com/commons-app/apps-android-commons/assets/99590/a28eebab-3986-442b-aa50-3e2375fc0891)

nicolas-raoul commented:
There seems to be something special with that pin, it is always grey even when I test with various `batchSize` values such as 5 or 1.

Also, the loading of pins sometimes stops for like 30 seconds before continuing. Maybe it could be related to some SPARQL queries staying in `Sending request` state as seen below?
![Screenshot_20240612-223840_HttpCanary](https://github.com/commons-app/apps-android-commons/assets/99590/968fe0a4-81ce-4091-9091-38b9c901e05f)

kanahia1 commented:
Hey @nicolas-raoul, Thanks for reviewing!
I have implemented pin changing color on clicking on it (if not loaded)

https://github.com/commons-app/apps-android-commons/assets/114223204/5cbc51e2-744f-4f5c-9400-ce3e4bad37cd

---

The pin is behaving really strangely 👻 as for one time it loads as red pin. But when I loaded again it became grey but on clicking it turns red which seems strange to me as we are using same query to load the pins in batch and loading details for the pins (if not loaded)



https://github.com/commons-app/apps-android-commons/assets/114223204/4334e13b-3bf5-40e5-9f7c-41408051e702


https://github.com/commons-app/apps-android-commons/assets/114223204/608207d2-8a66-44c7-abdb-289819934487




---

Could the delay be related with the server side as from the logs it can be seen that
HTTP request ended at `11:13:32.209`
then another GET request was started at `11:13:32.637`
but the data from the request was received at `11:14:23.618` which creates delay of `51 sec` but this seems to happen only once per running all the queries.

```
2024-06-13 11:13:32.209  8736-8783  OkHttp                  fr.free.nrw.commons.beta             V  <-- END HTTP (1820-byte body)
2024-06-13 11:13:32.576  8736-8736  Choreographer           fr.free.nrw.commons.beta             I  Skipped 41 frames!  The application may be doing too much work on its main thread.
2024-06-13 11:13:32.637  8736-8783  OkHttp                  fr.free.nrw.commons.beta             V  --> GET https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flabel%29%20AS%20%3Flabel%29%0A%20%20%28SAMPLE%28%3Fclass%29%20AS%20%3Fclass%29%0A%20%20%28SAMPLE%28%3Fdescription%29%20AS%20%3Fdescription%29%0A%20%20%28SAMPLE%28%3FclassLabel%29%20AS%20%3FclassLabel%29%0A%20%20%28SAMPLE%28%3Fpic%29%20AS%20%3Fpic%29%0A%20%20%28SAMPLE%28%3Fdestroyed%29%20AS%20%3Fdestroyed%29%0A%20%20%28SAMPLE%28%3FendTime%29%20AS%20%3FendTime%29%0A%20%20%28SAMPLE%28%3FwikipediaArticle%29%20AS%20%3FwikipediaArticle%29%0A%20%20%28SAMPLE%28%3FcommonsArticle%29%20AS%20%3FcommonsArticle%29%0A%20%20%28SAMPLE%28%3FcommonsCategory%29%20AS%20%3FcommonsCategory%29%0AWHERE%20%7B%0A%20%20SERVICE%20%3Chttps%3A%2F%2Fquery.wikidata.org%2Fsparql%3E%20%7B%0A%20%20%20%20values%20%3Fitem%20%7B%0A%20%20%20%20%20%20%0Awd%3AQ207581%0Awd%3AQ633002%0Awd%3AQ1777764%0A%20%20%20%20%7D%0A%20%20%7D%0A%0A%20%20%23%20Get%20the%20label%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20label%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%3Fitem%20rdfs%3Alabel%20%3FitemLabelPreferredLanguage.%20FILTER%20%28lang%28%3FitemLabelPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20rdfs%3Alabel%20%3FitemLabelAnyLanguage%7D%0A%20%20BIND%28COALESCE%28%3FitemLabelPreferredLanguage%2C%20%3FitemLabelAnyLanguage%2C%20%22%3F%22%29%20as%20%3Flabel%29%0A%0A%20%20%23%20Get%20the%20description%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20description%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%3Fitem%20schema%3Adescription%20%3FitemDescriptionPreferredLanguage.%20FILTER%20%28lang%28%3FitemDescriptionPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20schema%3Adescription%20%3FitemDescriptionAnyLanguage%7D%0A%20%20BIND%28COALESCE%28%3FitemDescriptionPreferredLanguage%2C%20%3FitemDescriptionAnyLanguage%2C%20%22%3F%22%29%20as%20%3Fdescription%29%0A%0A%20%20%23%20Get%20the%20class%20label%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20label%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%0A%20%20%3Fitem%20p%3AP31%2Fps%3AP31%20%3Fclass.%0A%20%20%20%20OPTIONAL%20%7B%3Fclass%20rdfs%3Alabel%20%3FclassLabelPreferredLanguage.%20FILTER%20%28lang%28%3FclassLabelPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20%20%20OPTIONAL%20%7B%3Fclass%20rdfs%3Alabel%20%3FclassLabelAnyLanguage%7D%0A%20%20%20%20BIND%28COALESCE%28%3FclassLabelPreferredLanguage%2C%20%3FclassLabelAnyLanguage%2C%20%22%3F%22%29%20as%20%3FclassLabel%29%0A%20%20%7D%0A%0A%20%20OPTIONAL%20%7B%0A%20%20%3Fitem%20p%3AP31%2Fps%3AP31%20%3Fclass.%0A%20%20%7D%0A%0A%20%20%23%20Get%20picture%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP18%20%3Fpic%7D%0A%0A%20%20%23%20Get%20existence%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP576%20%3Fdestroyed%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP582%20%3FendTime%7D%0A%0A%20%20%23%20Get%20Commons%20category%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP373%20%3FcommonsCategory%7D%0A%0A%20%20%23%20Get%20Wikipedia%20article%0A%20%20OPTIONAL%20%7B%0A%20%20%20%20%3FwikipediaArticle%20schema%3Aabout%20%3Fitem.%0A%20%20%20%20%3FwikipediaArticle%20schema%3AisPartOf%20%3Chttps%3A%2F%2Fen.wikipedia.org%2F%3E.%20%23%20TODO%20internationalization%0A%20%20%7D%0A%0A%20%20%23%20Get%20Commons%20article%0A%20%20OPTIONAL%20%7B%0A%20%20%20%20%3FcommonsArticle%20schema%3Aabout%20%3Fitem.%0A%20%20%20%20%3FcommonsArticle%20schema%3AisPartOf%20%3Chttps%3A%2F%2Fcommons.wikimedia.org%2F%3E.%0A%20%20%7D%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json
2024-06-13 11:13:32.637  8736-8783  OkHttp                  fr.free.nrw.commons.beta             V  --> END GET
2024-06-13 11:13:56.887  8736-8736  ViewRootIm...nActivity] fr.free.nrw.commons.beta             I  ViewPostIme pointer 0
2024-06-13 11:13:56.891  8736-8736  GestureDetector         fr.free.nrw.commons.beta             I  obtain mCurrentDownEvent. id: 771362179 caller: org.osmdroid.views.MapView.dispatchTouchEvent:1147 android.view.ViewGroup.dispatchTransformedTouchEvent:3328 android.view.ViewGroup.dispatchTouchEvent:2939 
2024-06-13 11:13:56.958  8736-8736  GestureDetector         fr.free.nrw.commons.beta             I  obtain mCurrentMotionEventRaw. action: 2 id: 808033211
2024-06-13 11:13:57.323  8736-8736  ViewRootIm...nActivity] fr.free.nrw.commons.beta             I  ViewPostIme pointer 1
2024-06-13 11:14:05.313  8736-8736  ViewRootIm...nActivity] fr.free.nrw.commons.beta             I  ViewPostIme pointer 0
2024-06-13 11:14:05.318  8736-8736  GestureDetector         fr.free.nrw.commons.beta             I  obtain mCurrentDownEvent. id: 684248822 caller: org.osmdroid.views.MapView.dispatchTouchEvent:1147 android.view.ViewGroup.dispatchTransformedTouchEvent:3328 android.view.ViewGroup.dispatchTouchEvent:2939 
2024-06-13 11:14:05.323  8736-8736  GestureDetector         fr.free.nrw.commons.beta             I  obtain mCurrentMotionEventRaw. action: 2 id: 635506627
2024-06-13 11:14:05.615  8736-8736  ViewRootIm...nActivity] fr.free.nrw.commons.beta             I  ViewPostIme pointer 1
2024-06-13 11:14:23.618  8736-8783  OkHttp                  fr.free.nrw.commons.beta             V  <-- 200 https://query.wikidata.org/sparql?query=SELECT%0A%20%20%3Fitem%0A%20%20%28SAMPLE%28%3Flabel%29%20AS%20%3Flabel%29%0A%20%20%28SAMPLE%28%3Fclass%29%20AS%20%3Fclass%29%0A%20%20%28SAMPLE%28%3Fdescription%29%20AS%20%3Fdescription%29%0A%20%20%28SAMPLE%28%3FclassLabel%29%20AS%20%3FclassLabel%29%0A%20%20%28SAMPLE%28%3Fpic%29%20AS%20%3Fpic%29%0A%20%20%28SAMPLE%28%3Fdestroyed%29%20AS%20%3Fdestroyed%29%0A%20%20%28SAMPLE%28%3FendTime%29%20AS%20%3FendTime%29%0A%20%20%28SAMPLE%28%3FwikipediaArticle%29%20AS%20%3FwikipediaArticle%29%0A%20%20%28SAMPLE%28%3FcommonsArticle%29%20AS%20%3FcommonsArticle%29%0A%20%20%28SAMPLE%28%3FcommonsCategory%29%20AS%20%3FcommonsCategory%29%0AWHERE%20%7B%0A%20%20SERVICE%20%3Chttps%3A%2F%2Fquery.wikidata.org%2Fsparql%3E%20%7B%0A%20%20%20%20values%20%3Fitem%20%7B%0A%20%20%20%20%20%20%0Awd%3AQ207581%0Awd%3AQ633002%0Awd%3AQ1777764%0A%20%20%20%20%7D%0A%20%20%7D%0A%0A%20%20%23%20Get%20the%20label%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20label%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%3Fitem%20rdfs%3Alabel%20%3FitemLabelPreferredLanguage.%20FILTER%20%28lang%28%3FitemLabelPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20rdfs%3Alabel%20%3FitemLabelAnyLanguage%7D%0A%20%20BIND%28COALESCE%28%3FitemLabelPreferredLanguage%2C%20%3FitemLabelAnyLanguage%2C%20%22%3F%22%29%20as%20%3Flabel%29%0A%0A%20%20%23%20Get%20the%20description%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20description%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%3Fitem%20schema%3Adescription%20%3FitemDescriptionPreferredLanguage.%20FILTER%20%28lang%28%3FitemDescriptionPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20schema%3Adescription%20%3FitemDescriptionAnyLanguage%7D%0A%20%20BIND%28COALESCE%28%3FitemDescriptionPreferredLanguage%2C%20%3FitemDescriptionAnyLanguage%2C%20%22%3F%22%29%20as%20%3Fdescription%29%0A%0A%20%20%23%20Get%20the%20class%20label%20in%20the%20preferred%20language%20of%20the%20user%2C%20or%20any%20other%20language%20if%20no%20label%20is%20available%20in%20that%20language.%0A%20%20OPTIONAL%20%7B%0A%20%20%3Fitem%20p%3AP31%2Fps%3AP31%20%3Fclass.%0A%20%20%20%20OPTIONAL%20%7B%3Fclass%20rdfs%3Alabel%20%3FclassLabelPreferredLanguage.%20FILTER%20%28lang%28%3FclassLabelPreferredLanguage%29%20%3D%20%22en%22%29%7D%0A%20%20%20%20OPTIONAL%20%7B%3Fclass%20rdfs%3Alabel%20%3FclassLabelAnyLanguage%7D%0A%20%20%20%20BIND%28COALESCE%28%3FclassLabelPreferredLanguage%2C%20%3FclassLabelAnyLanguage%2C%20%22%3F%22%29%20as%20%3FclassLabel%29%0A%20%20%7D%0A%0A%20%20OPTIONAL%20%7B%0A%20%20%3Fitem%20p%3AP31%2Fps%3AP31%20%3Fclass.%0A%20%20%7D%0A%0A%20%20%23%20Get%20picture%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP18%20%3Fpic%7D%0A%0A%20%20%23%20Get%20existence%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP576%20%3Fdestroyed%7D%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP582%20%3FendTime%7D%0A%0A%20%20%23%20Get%20Commons%20category%0A%20%20OPTIONAL%20%7B%3Fitem%20wdt%3AP373%20%3FcommonsCategory%7D%0A%0A%20%20%23%20Get%20Wikipedia%20article%0A%20%20OPTIONAL%20%7B%0A%20%20%20%20%3FwikipediaArticle%20schema%3Aabout%20%3Fitem.%0A%20%20%20%20%3FwikipediaArticle%20schema%3AisPartOf%20%3Chttps%3A%2F%2Fen.wikipedia.org%2F%3E.%20%23%20TODO%20internationalization%0A%20%20%7D%0A%0A%20%20%23%20Get%20Commons%20article%0A%20%20OPTIONAL%20%7B%0A%20%20%20%20%3FcommonsArticle%20schema%3Aabout%20%3Fitem.%0A%20%20%20%20%3FcommonsArticle%20schema%3AisPartOf%20%3Chttps%3A%2F%2Fcommons.wikimedia.org%2F%3E.%0A%20%20%7D%0A%7D%0AGROUP%20BY%20%3Fitem%0A&format=json (50980ms)

```

kanahia1 commented:
https://github.com/commons-app/apps-android-commons/assets/114223204/388ab83f-52ff-4388-90eb-6f44ce4c3ae7

Removed search this area button, _needs to work on caching though_

nicolas-raoul commented:
Great! 🙂

kanahia1 commented:
I have implemented caching and loading speed is reduced 

**Without Caching**

https://github.com/commons-app/apps-android-commons/assets/114223204/c034db60-b0ad-47be-9075-d8b515158201

**With Caching**

https://github.com/commons-app/apps-android-commons/assets/114223204/57d7c792-ef0a-4024-b8c9-df73c189c182

kanahia1 commented on app/src/main/java/fr/free/nrw/commons/mwapi/OkHttpJsonApiClient.java:
```diff
@@ -397,6 +397,82 @@ public List<Place> getNearbyPlaces(
+    }
+
+    @Nullable
+    public List<Place> getPlaces(
```
Hi @nicolas-raoul, I have factorized methods in newer in commit https://github.com/commons-app/apps-android-commons/pull/5731/commits/438aa3b41b107b4776c5b4ac334b3d283796c204

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/PlacesRepository.java:
```diff
@@ -0,0 +1,25 @@
+        return localDataSource.savePlace(place);
+    }
+
+    public Place fetchPlace(LatLng latLng){
```
Several Nearby pins (Wikidata items) can have the exact same latitude/longitude, it actually happens rather often.

Wouldn't it work better in such cases if the cache indexed on QID rather than on LatLng?

kanahia1 replied:
Sure, done in the newer commit https://github.com/commons-app/apps-android-commons/pull/5731/commits/a28117397d1dcb5e6d5b56b970727dde11ff5074

nicolas-raoul commented:
Would you mind fixing the conflict? :-)

nicolas-raoul commented:
After installing from zero, or aftdr deleting cache/data, pins do not load until I move the map:

https://github.com/commons-app/apps-android-commons/assets/99590/4b0b953f-f59e-426e-b4a7-fe91d153afbb

kanahia1 commented:
@nicolas-raoul Looking into the issue!

kanahia1 commented:
Hi @nicolas-raoul, I have fixed the bug in  https://github.com/commons-app/apps-android-commons/pull/5731/commits/2251c62355e021d7f383836cf9a984ff13ff11a9


https://github.com/commons-app/apps-android-commons/assets/114223204/2e4f8632-2da0-4547-b60c-1afa4b49b12d

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -1861,46 +1943,52 @@ public boolean onItemLongPress(int index, OverlayItem item) {
-                @Override
-                public boolean onItemSingleTapUp(int index, OverlayItem item) {
-                    final Place place = nearbyBaseMarkers.get(index).getPlace();
+            Marker marker = new Marker(binding.map);
```
Note: I created https://github.com/commons-app/apps-android-commons/issues/5780 about factorization, not your fault, hopefully someone else will factorize later.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/nearby/presenter/NearbyParentFragmentPresenter.java:
```diff
@@ -327,17 +328,12 @@ public void searchViewGainedFocus() {
-                updateMapAndList(SEARCH_CUSTOM_AREA);
-            }
-        };
+    public void searchInTheArea(){
```
javadoc

nicolas-raoul reviewed (changes requested):
When I load Nearby for the first time, it loads a much wider rectangle than what is seen on screen.
Then when I move the map and let the pins refresh, that time only pins visible on screen are loaded.

Any idea what makes it load too many pins the first time? It makes the initial query longer and makes many unneeded queries to the server.

kanahia1 commented:
> When I load Nearby for the first time, it loads a much wider rectangle than what is seen on screen. Then when I move the map and let the pins refresh, that time only pins visible on screen are loaded.
> 
> Any idea what makes it load too many pins the first time? It makes the initial query longer and makes many unneeded queries to the server.

Working on this issue ⚒️

kanahia1 commented:
Hello @nicolas-raoul, I have added  jcenter in https://github.com/commons-app/apps-android-commons/pull/5731/commits/4f7cbfc741193b2aba6cb1e6364e9d7c63faa2ea which seems to be fix for https://github.com/commons-app/apps-android-commons/actions/runs/10131832421/job/28015018616?pr=5731 

> Could not resolve all files for configuration ':app:betaDebugCompileClasspath'.
   > Could not find com.dinuscxj:circleprogressbar:1.1.1. 

as this library is hosted on jcenter https://mvnrepository.com/artifact/com.dinuscxj/circleprogressbar?repo=jcenter

nicolas-raoul commented:
Maybe not a big issue, but I see the pin at the top appearing and disappearing:

https://github.com/user-attachments/assets/5a78665b-0b00-451e-ad04-6c380b9380c0

sorry for the bad video quality

kanahia1 commented:
> Maybe not a big issue, but I see the pin at the top appearing and disappearing:
> 
>  PXL_20240730_012259644.mp4 
> sorry for the bad video quality

If possible can you please share the location over zulip (for testing)

nicolas-raoul reviewed (changes requested):
Good job, you are quickly approaching mergeability for this pull request :-)

kanahia1 commented on app/src/main/java/fr/free/nrw/commons/nearby/fragments/NearbyParentFragment.java:
```diff
@@ -185,19 +198,25 @@ public class NearbyParentFragment extends CommonsDaggerSupportFragment
     private Place nearestPlace;
+    private volatile boolean stopQuery;
+
+    private List<Place> updatedPlaceList;
```
Thanks for reviewing, I have made all suggested changes in the newer commit.

nicolas-raoul commented:
I tested more there are still a few bugs but soon mergeable I think. 🙂

The biggest problem: It seems like the first time (for instance after installing the app) Nearby pins do not load, until I switch to home activity then come back:

https://github.com/user-attachments/assets/8739fad9-d86f-4992-b172-850c93c795aa

kanahia1 commented:
I have shared a newer commit fixing this bug. Can you please test it.

https://github.com/user-attachments/assets/f5207595-cd7f-4134-aa0f-8f96a73ae500

nicolas-raoul commented:
Fantastic, thanks Kanahia for this pull request, it is a great improvement!
I will merge for now, and submit anything else as new GitHub issues. 🙂
